File: chapter1/exercises/comparing_means.py
# RCDS Introduction to probability and statistical inference.
# Jesús Urtasun Elizari. ICL 2024 / 2025.
# Chapter 1. Parameter estimation and hypotesis testing.

# Import libraries
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats

# Random seed
np.random.seed(0)

# Comparing means .............................................................
print("Comparing means")

# Generate two sets of gaussian observations
sample1 = np.random.normal(loc = 50, scale = 5, size = 30)
sample2 = np.random.normal(loc = 55, scale = 5, size = 30)

# Plot observations as histograms
plt.hist(sample1, bins = 15, alpha = 0.5, color = "blue", label = "Sample 1")
plt.hist(sample2, bins = 15, alpha = 0.5, color = "red", label = "Sample 2")
plt.xlabel("Value")
plt.ylabel("Frequency")
plt.title("Histogram of gaussian observations")
plt.legend()
plt.show()

# Calculate t-statistic and p-value manually ..................................

# Calculate the sample means and standard deviations
mean1, mean2 = np.mean(sample1), np.mean(sample2)
std1, std2 = np.std(sample1, ddof = 1), np.std(sample2, ddof = 1)
n1, n2 = len(sample1), len(sample2)

# Calculate the standard error of the difference
se = np.sqrt((std1 ** 2 / n1) + (std2 ** 2 / n2))

# Calculate the t-statistic
t_stat_manual = (mean1 - mean2) / se

# Calculate degrees of freedom
df = n1 + n2 - 2

# Calculate two-tailed statistic p-value
p_value_manual = 2 * (1 - stats.t.cdf(abs(t_stat_manual), df = df))
print(f"Manual calculation:")
print(f"t-statistic = {t_stat_manual:.4f}")
print(f"p-value = {p_value_manual:.4f}")

# Calculate t-statistic and p-value with scipy ................................

# Calculate two-tailed statistic and p-value
t_stat_lib, p_value_lib = stats.ttest_ind(sample1, sample2)
print(f"\nUsing SciPy library:")
print(f"t-statistic = {t_stat_lib:.4f}")
print(f"p-value = {p_value_lib:.4f}")

File: chapter1/exercises/comparing_means.R
# RCDS Further hypothesis testing
# Jesús Urtasun Elizari - Imperial College London
# Chapter 1 - Comparing means

# Chapter 1. Parameter estimation, comparing means and variances

# Install libraries
library(tidyverse)

# Set working directory
setwd("/Users/jurtasun/Desktop/Courses/ICL/RCDS_Further_hypothesis_testing/chapter1")

# Question: do types 4 and 5 have the same mean luminosity?
# Question: do types 2 and 3 have the same mean luminosity?

# Read input data
data <- read_csv("stars.csv")
type_key <- c('Brown Dwarf', 'Red Dwarf', 'White Dwarf', 'Main Sequence', 'Supergiant','Hypergiant')
spectral_classes <- c('O', 'B', 'A', 'F', 'G', 'K', 'M')
      
# Explore data
head(data)
class(data)
dim(data)

# Store type and spectral class as factors
data$type <- factor(data$type)
data$spectral_class <- factor(data$spectral_class, levels=spectral_classes)

# Plot type 4 (supergiant) and type 5 (hypergiant)
data %>% filter(type %in% c(4,5)) %>%
      ggplot(aes(x = log(luminosity), fill = type)) + 
      geom_histogram(alpha = 0.5, bins = 10) + 
      guides(fill = "none") +
      theme_classic() + 
      facet_wrap(~ type, ncol = 1)

# Formulate null and alternative hypothesis
# H0: they come from same distribution
# H1: they come from different distributions

# Get log luminosity and compute mean values
table(data$type)
type4 <- data %>% filter(type == 4) %>% pull(luminosity) %>% log
type5 <- data %>% filter(type == 5) %>% pull(luminosity) %>% log
mean4 <- mean(type4)
mean5 <- mean(type5)
print(paste('Type 4:', mean4))
print(paste('Type 5:', mean5))
print(paste('difference:', mean4 - mean5))

# Mean values are similar, but is the difference between them statistically significant?
      
# Compute t-test
t.test(type4, type5, var.equal = TRUE, paired = FALSE)

# Store obtained t value and d.o.f
t_obs <- t.test(type4, type5, var.equal = TRUE, paired = FALSE)$statistic
df <- length(type4) + length(type5) - 2
print(paste("degrees of freedom:", df))

# Build range for t distribution
tmin <- -4
tmax <- 4
x <- seq(tmin, tmax, 0.01)
plot(x, dt(x, df), xlab = "t", ylab = "pdf", type = "l", col = "grey")

# The area of the shaded region is the two-tailed p-value
lower_tail <- seq(tmin, -t_obs, 0.01)
upper_tail <- seq(t_obs, tmax, 0.01)
polygon(c(lower_tail, -t_obs, tmin), c(dt(lower_tail, df), 0, 0), border = NA, col = "lightgrey")
polygon(c(upper_tail, tmax, t_obs), c(dt(upper_tail, df), 0, 0), border = NA, col = "lightgrey")

